### PR TITLE
feat: interim care order (FPLA-981)

### PR DESCRIPTION
**Pull request by joez-f**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-981

### Change description ###

Add a choice in the 'Create an order' event flow so that users may specify the type of care order they want to create (interim or final). This then updates the file name and document accordingly

Test this as part of #803

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerTest.java [hunk 1]
```diff
@@ -157,66 +160,79 @@ private CallbackRequest buildCallbackRequest() {
...
                 .build();
 
-            aboutToSubmitAssertions(callbackResponse.getData(), expectedCareOrder);
+            aboutToSubmitAssertions(callbackResponse.getData(), expectedOrder);
         }
 
         private CaseDetails buildCaseDetails(CaseData.CaseDataBuilder builder) {
             return CaseDetails.builder()
-                .data(mapper.convertValue(builder.build(), new TypeReference<>() {}))
+                .data(mapper.convertValue(builder.build(), new TypeReference<>() {
```
Comment (Jamesnelson117) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerTest.java]: I think accidental change

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerTest.java [hunk 2]
```diff
@@ -315,8 +334,16 @@ private CaseDetails generateBlankOrderCaseDetails() {
...
-            final CaseData.CaseDataBuilder dataBuilder = generateCommonCareOrderDetails();
+        private CaseDetails generateCareOrderCaseDetailsWithoutFurtherDirections() {
+            final CaseData.CaseDataBuilder dataBuilder = generateCommonOrderDetails(CARE_ORDER, INTERIM);
+
+            return CaseDetails.builder()
+                .data(mapper.convertValue(dataBuilder.build(), new TypeReference<>() {}))
+                .build();
+        }
+
+        private CaseDetails generateCareOrderCaseDetailsWithInterimSubtype() {
```
Comment (Jamesnelson117) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerTest.java]: Very similar to 'generateCareOrderCaseDetailsWithFinalSubtype'. Can we refactor into the one method that takes in subType as param?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/GeneratedOrderType.java [hunk 3]
```diff
@@ -1,16 +1,26 @@
...
 
     private final String label;
 
-    GeneratedOrderType(String label) {
-        this.label = label;
+    public String getFullType() {
+        return getFullType(null);
+    }
+
+    public String getFullType(GeneratedOrderSubtype subtype) {
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/GeneratedOrderType.java]: Maybe better to be place on the model?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java [hunk 4]
```diff
@@ -183,7 +201,14 @@ private String getLocalAuthorityName(String caseLocalAuthority) {
         return localAuthorityNameLookupConfiguration.getLocalAuthorityName(caseLocalAuthority);
     }
 
-    private String supervisionOrderDetails(int numOfChildren, String caseLocalAuthority, int numOfMonths) {
+    private String careOrderDetails(int numOfChildren, String caseLocalAuthority, OrderTypeAndDocument typeAndDoc) {
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java]: getFormattedCareOrderDetails or similar is a better name

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java [hunk 5]
```diff
@@ -158,8 +173,11 @@ public OrderTypeAndDocument buildOrderTypeAndDocument(OrderTypeAndDocument typeA
         return orderTemplateBuilder.build();
     }
 
-    public String generateOrderDocumentFileName(String type) {
-        return type.toLowerCase().replaceAll("[()]", "").replaceAll("[ ]", "_") + ".pdf";
+    public String generateOrderDocumentFileName(GeneratedOrderType orderType, GeneratedOrderSubtype orderSubtype) {
+        String type = orderType.getLabel().toLowerCase().replace("(", "").replace(")", "").replace(" ", "_");
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java]: This is getting a bit messy, consider adding the type to the enum

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerTest.java [hunk 6]
```diff
@@ -328,31 +355,25 @@ private CaseDetails generateCareOrderCaseDetailsWithFurtherDirections() {
...
-
         private CaseDetails generateSupervisionOrderCaseDetails() {
-            final CaseData.CaseDataBuilder dataBuilder = CaseData.builder()
-                .orderTypeAndDocument(OrderTypeAndDocument.builder().type(SUPERVISION_ORDER).build())
-                .orderFurtherDirections(FurtherDirections.builder()
-                    .directionsNeeded("No")
-                    .build())
+            final CaseData.CaseDataBuilder dataBuilder = generateCommonOrderDetails(SUPERVISION_ORDER, null);
+
+            dataBuilder.orderFurtherDirections(FurtherDirections.builder()
```
[the remark this answers is not in the document]
Comment (joez-f) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerTest.java]: Sure

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerTest.java (hunk 1 above)
Comment (joez-f) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerTest.java]: yes, my auto formatter - will fix

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerTest.java (hunk 2 above)
Comment (joez-f) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerTest.java]: yep 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/GeneratedOrderType.java (hunk 3 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/enums/GeneratedOrderType.java]: will do

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java (hunk 5 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java]: will add type as a value to enum

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/OrderTypeAndDocument.java [hunk 7]
```diff
@@ -2,13 +2,29 @@
...
+    public String getFullType() {
+        return getFullType(null);
+    }
+
+    public String getFullType(GeneratedOrderSubtype subtype) {
+        return (subtype != null) ? subtype.getLabel() + " " + this.type.getLabel().toLowerCase() : this.type.getLabel();
+    }
+
+    public boolean hasInterimSubtype() {
+        return (subtype == INTERIM);
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/OrderTypeAndDocument.java]:
```suggestion
        return subtype == INTERIM;
```

**Pull request merged**